assets/src/features/devices/useUserMedia.tsx

```diff
@@ -0,0 +1,451 @@
+
+const isAnyDeviceDifferentFromLastSession = (
+  lastVideoDevice: MediaDeviceInfo | null,
+  lastAudioDevice: MediaDeviceInfo | null,
+  currentDevices: CurrentDevices | null
+) =>
+  isDeviceDifferentFromLastSession(lastVideoDevice, currentDevices?.videoinput || null) ||
+  isDeviceDifferentFromLastSession(lastAudioDevice, currentDevices?.audioinput || null);
+
+const stopTracks = (requestedDevices: MediaStream) => {
+  requestedDevices.getTracks().forEach((track) => {
+    track.stop();
+  });
+};
+
+const getExactConstraints = (
+  shouldAskForVideo: boolean,
+  videoConstraints: MediaTrackConstraints | undefined,
+  previousVideoDevice: MediaDeviceInfo | null,
+  shouldAskForAudio: boolean,
+  audioConstraints: MediaTrackConstraints | undefined,
+  previousAudioDevice: MediaDeviceInfo | null
+): MediaStreamConstraints => ({
+  video: shouldAskForVideo && {
+    ...videoConstraints,
+    deviceId: { exact: previousVideoDevice?.deviceId },
+  },
+  audio: shouldAskForAudio && {
+    ...audioConstraints,
+    deviceId: { exact: previousAudioDevice?.deviceId },
+  },
+});
+
+const getExactConstraintsIfPossible = (
+  shouldAskForVideo: boolean,
+  videoIdToStart: string | undefined,
+  videoConstraints: MediaTrackConstraints | undefined,
+  shouldAskForAudio: boolean,
+  audioIdToStart: string | undefined,
+  audioConstraints: MediaTrackConstraints | undefined
+) => ({
+  video:
+    shouldAskForVideo && !!videoIdToStart
+      ? {
+          ...videoConstraints,
+          deviceId: { exact: videoIdToStart },
+        }
+      : videoConstraints,
+  audio:
+    shouldAskForAudio && !!audioIdToStart
+      ? {
+          ...audioConstraints,
+          deviceId: { exact: audioIdToStart },
+        }
+      : audioConstraints,
+});
+
+const REQUESTING = { type: "Requesting" } as const;
+const NOT_REQUESTED = { type: "Not requested" } as const;
+
+const shouldAskForLastDevices = (
+  previousVideoDevice: MediaDeviceInfo | null,
+  previousAudioDevice: MediaDeviceInfo | null
+) => previousVideoDevice?.deviceId || previousAudioDevice?.deviceId;
+
+const shouldAskForAnyDevice = (requestedDevices: null | MediaStream) => requestedDevices === null;
+
+/**
+ * Hook that returns the list of available devices
+ */
+export const useUserMedia = ({
+  getLastAudioDevice,
+  saveLastAudioDevice,
+  getLastVideoDevice,
+  saveLastVideoDevice,
+  videoTrackConstraints,
+  audioTrackConstraints,
+  refetchOnMount,
+}: UseUserMediaConfig): UseUserMedia => {
+  const [state, setState] = useState<UseUserMediaState>({
+    video: { type: "Not requested" },
+    audio: { type: "Not requested" },
+    audioMedia: null,
+    videoMedia: null,
+    audioError: null,
+    videoError: null,
+  });
+
+  const audioConstraints = useMemo(() => toMediaTrackConstraints(audioTrackConstraints), [audioTrackConstraints]);
+  const videoConstraints = useMemo(() => toMediaTrackConstraints(videoTrackConstraints), [videoTrackConstraints]);
+
+  const skip = useRef<boolean>(false);
```

Comment:
From https://react.dev/reference/react/useRef
> useRef is a React Hook that lets you reference a value that’s not needed for rendering.

In this case I want to guard `init` method from additional invocations during rerender, so I used this mutable variable because changing this variable does not invoke rerender